madmom/features/key.py

```diff
@@ -0,0 +1,144 @@
+from ..processors import SequentialProcessor
+
+
+KEY_LABELS = ['A major', 'Bb major', 'B major', 'C major', 'Db major',
+              'D major', 'Eb major', 'E major', 'F major', 'F# major',
+              'G major', 'Ab major', 'A minor', 'Bb minor', 'B minor',
+              'C minor', 'C# minor', 'D minor', 'D# minor', 'E minor',
+              'F minor', 'F# minor', 'G minor', 'G# minor']
+
+
+def key_prediction_to_label(prediction):
+    """
+    Convert key class id to a human-readable key name.
+
+    Parameters
+    ----------
+    prediction : numpy array
+        Array containing the probabilities of each key class.
+
+    Returns
+    -------
+    str
+        Human-readable key name.
+
+    """
+    return KEY_LABELS[prediction[0].argmax()]
+
+
+def write_key(key, filename):
+    """
+    Write key string to a file.
+
+    Parameters
+    ----------
+    key : str
+        Key name.
+    filename : str
+        Output filename.
+
+    Returns
+    -------
+    str
+        Key name.
+
+    """
+    open(filename, 'w').write(key)
+    return key
+
+
+class CNNKeyRecognitionProcessor(SequentialProcessor):
+    """
+    Recognise the global key of a musical piece using a Convolutional Neural
+    Network as described in [1]_.
+
+    Parameters
+    ----------
+    nn_files : list, optional
+        List with trained CNN model files. Per default ('None'), an ensemble
+        of networks will be used.
+    single_net : bool, optional
+        Use only a single CNN for prediction. This speeds up processing, but
+        slightly worsens the results.
+
+    References
+    ----------
+    .. [1] Filip Korzeniowski and Gerhard Widmer,
+           "End-to-End Musical Key Estimation Using a Convolutional Neural
+           Network", In Proceedings of the 25th European Signal Processing
+           Conference (EUSIPCO), Kos, Greece, 2017.
+           
+    Examples
+    --------
+    Create a CNNKeyRecognitionProcessor and pass a file through it.
+    The returned array represents the probability of each key class.
+    
+    >>> proc = CNNKeyRecognitionProcessor()
+    >>> proc  # doctest: +ELLIPSIS
+    <madmom.features.key.CNNKeyRecognitionProcessor object at 0x...>
+    >>> proc('tests/data/audio/sample.wav')  # doctest: +ELLIPSIS
+    array([[ 0.00096,  0.00226,  0.0169 ,  0.00138,  0.01767,  0.00352, ...,
+             0.17846,  0.12905,  0.06041,  0.08945,  0.00312,  0.08681]], dtype=float32)
+             
+    """
+ 
+    SINGLE_NET = False
+
+    def __init__(self, nn_files=None, single_net=SINGLE_NET, **kwargs):
```

Comment:
I don't like the `single_net` parameter, maybe a more generic `num_models` to state the maximum numbers of models to be evaluated would be more flexible. This could be moved to `NeuralNetworkEnsemble`, then all other programs/algorithms would benefit from it.

We should consider using fewer models in general in order to keep the package smaller. Supplying multiple networks is often not really needed for good performance and are often only supplied to experiment setups. But maybe this discussion should be moved somewhere else.

Reply:
Yes, but wouldn't it be weird to pass a long list of models to the `NeuralNetworkEnsemble`, and also pass a parameter that sets how many to select? You could just pass the shorter list to the ensemble in the first place.

I agree with the number of models. In this case, however, using the ensemble gives +3pp for Pop/Rock music, so I thought it's worth it.

Reply:
What I basically wanted to say that this option should go to a more central place if there's a need for it. Following your argumentation, `single_net` is not the way to go then ;)

Regarding the number of models: ensembles usually give a performance boost, that's why I also check in multiple models. I'm fine with 4 models to be checked in into madmom/models, so I think we can defer this discussion and aim for a global solution since this doesn't affect only key estimation.